### PR TITLE
feat(datagrid): disable and hide open detail button 

### PR DESCRIPTION
**Pull request by valentin-mladenov**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] If applicable, have a visual design approval

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## What is the current behavior?
Opening details on a datagrid is always possible.

Issue Number: CDE-1073

## What is the new behavior?
Datagrid row details can be disabled.
Datagrid row open details button can be disabled.
Datagrid row open details button can be hidden.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Review by dtsanevmw — COMMENTED** [review]
Check comments as discussed offline.

File: .storybook/stories/datagrid/datagrid-detail.stories.ts [hunk 1]
```diff
@@ -31,6 +31,10 @@ export default {
     clrDetailAriaLabelledBy: {
       description: "Id or multiple space separated Id's referencing existing text on the page",
     },
+    disabledElement: {
+      description: 'Disabled element page index.',
+      control: { type: 'number', min: -1, max: 50 },
+    },
```
[the remark this answers is not in the document]
Comment (valentin-mladenov) [reply .storybook/stories/datagrid/datagrid-detail.stories.ts]: Corrected in commit 56fa811

File: .storybook/stories/datagrid/datagrid-detail.stories.ts [hunk 2]
```diff
@@ -48,6 +52,7 @@ export default {
     compact: false,
     hidableColumns: false,
     height: 0,
+    disabledElement: -1,
```
[the remark this answers is not in the document]
Comment (valentin-mladenov) [reply .storybook/stories/datagrid/datagrid-detail.stories.ts]: Corrected in commit 56fa811

File: .storybook/stories/datagrid/datagrid-detail.stories.ts [hunk 3]
```diff
@@ -100,7 +105,11 @@ const DetailTemplate: StoryFn = args => {
           <ng-container ${args.hidableColumns ? '*clrDgHideableColumn' : ''}>Electronegativity</ng-container>
         </clr-dg-column>
 
-        <clr-dg-row *clrDgItems="let element of elements; let index = index" [clrDgItem]="element">
+        <clr-dg-row
+          *clrDgItems="let element of elements; let index = index"
+          [clrDgItem]="element"
+          [clrDgDetailsDisabled]="disabledElement === index"
```
[the remark this answers is not in the document]
Comment (valentin-mladenov) [reply .storybook/stories/datagrid/datagrid-detail.stories.ts]: Corrected in commit 56fa811

File: projects/angular/clarity.api.md [hunk 4]
```diff
@@ -1766,6 +1766,8 @@ export class ClrDatagridRow<T = any> implements AfterContentInit, AfterViewInit
     // (undocumented)
     detailButton: ElementRef<HTMLButtonElement>;
     // (undocumented)
+    detailsDisabled: boolean;
```
[the remark this answers is not in the document]
Comment (valentin-mladenov) [reply projects/angular/clarity.api.md]: Corrected in commit 56fa811

File: projects/angular/clarity.api.md [hunk 5]
```diff
@@ -1828,7 +1830,7 @@ export class ClrDatagridRow<T = any> implements AfterContentInit, AfterViewInit
     // (undocumented)
     get _view(): any;
     // (undocumented)
-    static ɵcmp: i0.ɵɵComponentDeclaration<ClrDatagridRow<any>, "clr-dg-row", never, { "item": "clrDgItem"; "clrDgSelectable": "clrDgSelectable"; "selected": "clrDgSelected"; "expanded": "clrDgExpanded"; "clrDgDetailOpenLabel": "clrDgDetailOpenLabel"; "clrDgDetailCloseLabel": "clrDgDetailCloseLabel"; "clrDgRowSelectionLabel": "clrDgRowSelectionLabel"; }, { "selectedChanged": "clrDgSelectedChange"; "expandedChange": "clrDgExpandedChange"; }, ["dgCells"], ["clr-dg-row-detail", "clr-dg-action-overflow", "clr-dg-cell"], false, never>;
+    static ɵcmp: i0.ɵɵComponentDeclaration<ClrDatagridRow<any>, "clr-dg-row", never, { "detailsDisabled": "clrDgDetailsDisabled"; "item": "clrDgItem"; "clrDgSelectable": "clrDgSelectable"; "selected": "clrDgSelected"; "expanded": "clrDgExpanded"; "clrDgDetailOpenLabel": "clrDgDetailOpenLabel"; "clrDgDetailCloseLabel": "clrDgDetailCloseLabel"; "clrDgRowSelectionLabel": "clrDgRowSelectionLabel"; }, { "selectedChanged": "clrDgSelectedChange"; "expandedChange": "clrDgExpandedChange"; }, ["dgCells"], ["clr-dg-row-detail", "clr-dg-action-overflow", "clr-dg-cell"], false, never>;
```
[the remark this answers is not in the document]
Comment (valentin-mladenov) [reply projects/angular/clarity.api.md]: Corrected in commit 56fa811

File: projects/angular/src/data/datagrid/datagrid-row.html [hunk 6]
```diff
@@ -124,6 +124,7 @@
         <div
           *ngIf="detailService.enabled"
           class="datagrid-detail-caret datagrid-fixed-column datagrid-cell"
+          [attr.aria-disabled]="detailsDisabled"
```
[the remark this answers is not in the document]
Comment (valentin-mladenov) [reply projects/angular/src/data/datagrid/datagrid-row.html]: Corrected in commit 56fa811

File: projects/angular/src/data/datagrid/datagrid-row.html [hunk 7]
```diff
@@ -137,6 +138,7 @@
             [attr.aria-expanded]="detailService.isRowOpen(item)"
             [attr.aria-controls]="detailService.id"
             aria-haspopup="dialog"
+            [disabled]="detailsDisabled"
```
[the remark this answers is not in the document]
Comment (valentin-mladenov) [reply projects/angular/src/data/datagrid/datagrid-row.html]: Corrected in commit 56fa811

File: projects/angular/src/data/datagrid/datagrid-row.ts [hunk 8]
```diff
@@ -138,7 +139,12 @@ export class ClrDatagridRow<T = any> implements AfterContentInit, AfterViewInit
             this.renderer.removeClass(this.el.nativeElement, 'datagrid-row-replaced');
           }
         }
-      )
+      ),
+      this.detailService.stateChange.subscribe(() => {
+        if (this.detailsDisabled && this.detailService.isRowOpen(this.item)) {
+          this.detailService.close();
+        }
+      })
```
[the remark this answers is not in the document]
Comment (valentin-mladenov) [reply projects/angular/src/data/datagrid/datagrid-row.ts]: Removed in commit 56fa811

File: projects/angular/src/data/datagrid/datagrid-row.spec.ts [hunk 9]
```diff
@@ -543,6 +543,123 @@ export default function (): void {
...
+    describe('Details', function () {
+      let context: TestContext<ClrDatagridRow<Item>, DatagridWithDisabledOrHiddenDetails>;
+
+      beforeEach(function () {
+        context = this.create(ClrDatagridRow, DatagridWithDisabledOrHiddenDetails, DATAGRID_SPEC_PROVIDERS);
+        context.detectChanges();
+      });
+
+      it('displays a clickable double caret when the row have details', function () {
+        expect(context.clarityElement.querySelector('button cds-icon[shape^=angle-double]')).not.toBeNull();
```
Comment (dtsanevmw) [inline projects/angular/src/data/datagrid/datagrid-row.spec.ts:556]: Can we just check for the actual button?
```suggestion
        expect(context.clarityElement.querySelector('button.datagrid-detail-caret-button')).not.toBeNull();
```

File: projects/angular/src/data/datagrid/datagrid-row.spec.ts [hunk 10]
```diff
@@ -543,6 +543,123 @@ export default function (): void {
...
+        context.testComponent.preState = context.testComponent.items[0];
+        tick();
+        context.detectChanges();
+
+        const button = context.clarityElement.querySelector('button.datagrid-detail-caret-button');
+        const detailBody = document.querySelector('div.clr-dg-detail-body-wrapper') as HTMLElement;
+
+        expect(button).not.toBeNull();
+        expect(button.querySelector('cds-icon[shape^=angle-double][direction^=left]')).not.toBeNull();
+        expect(detailBody.innerText).toBe('' + context.testComponent.preState.id);
```
Comment (dtsanevmw) [inline projects/angular/src/data/datagrid/datagrid-row.spec.ts:575]: Do we need to check for the button and icon again if we have that in the test above?

File: projects/angular/src/data/datagrid/datagrid-row.html [hunk 11]
```diff
@@ -132,11 +132,13 @@
             type="button"
             #detailButton
             class="datagrid-detail-caret-button"
+            aria-haspopup="dialog"
+            [disabled]="detailDisabled"
+            [class.is-not-visible]="detailHidden"
```
[the remark this answers is not in the document]
Comment (valentin-mladenov) [reply projects/angular/src/data/datagrid/datagrid-row.html]: Using `ngIf` in commit 00e868d

File: projects/angular/src/data/datagrid/datagrid-row.spec.ts (hunk 10 above)
Comment (valentin-mladenov) [reply projects/angular/src/data/datagrid/datagrid-row.spec.ts]: Corrected in commit 00e868d

File: projects/angular/src/data/datagrid/datagrid-row.spec.ts (hunk 9 above)
Comment (valentin-mladenov) [reply projects/angular/src/data/datagrid/datagrid-row.spec.ts]: Corrected in commit 00e868d

File: projects/angular/src/data/datagrid/datagrid-row.html [hunk 12]
```diff
@@ -132,11 +132,13 @@
...
             #detailButton
             class="datagrid-detail-caret-button"
+            aria-haspopup="dialog"
+            [disabled]="detailDisabled"
+            *ngIf="!detailHidden"
             [class.is-open]="detailService.isRowOpen(item)"
             [attr.aria-label]="detailService.isRowOpen(item) ? clrDgDetailCloseLabel : clrDgDetailOpenLabel"
             [attr.aria-expanded]="detailService.isRowOpen(item)"
             [attr.aria-controls]="detailService.id"
-            aria-haspopup="dialog"
```
Comment (kevinbuhmann) [inline projects/angular/src/data/datagrid/datagrid-row.html:139]: I think this should be moved back. The aria attributes were groped together.

File: projects/demo/src/app/datagrid/detail/detail.html [hunk 13]
```diff
@@ -246,10 +246,12 @@ <h3>Fixed Height</h3>
   <clr-dg-column [clrDgField]="'color'"><ng-container *clrDgHideableColumn>Favorite color</ng-container></clr-dg-column>
 
   <clr-dg-row
-    *clrDgItems="let user of users | slice:0:4"
+    *clrDgItems="let user of users | slice:0:4; let i = index;"
     [clrDgItem]="user"
     [clrDgDetailOpenLabel]="'Open row details for ' + user.name"
     [clrDgDetailCloseLabel]="'Open row details for ' + user.name"
+    [clrDgDetailDisabled]="i===0"
```
Comment (kevinbuhmann) [inline projects/demo/src/app/datagrid/detail/detail.html:253]:
```suggestion
    [clrDgDetailDisabled]="i === 0"
```

File: projects/demo/src/app/datagrid/detail/detail.html [hunk 14]
```diff
@@ -246,10 +246,12 @@ <h3>Fixed Height</h3>
...
+    [clrDgDetailHidden]="i===1"
```
Comment (kevinbuhmann) [inline projects/demo/src/app/datagrid/detail/detail.html:254]:
```suggestion
    [clrDgDetailHidden]="i === 1"
```

File: projects/demo/src/app/datagrid/detail/detail.ts [hunk 15]
```diff
@@ -20,18 +20,26 @@ export class DatagridDetailDemo implements AfterViewInit {
...
   constructor(inventory: Inventory, private cdr: ChangeDetectorRef) {
     inventory.size = 103;
     inventory.reset();
     this.users = inventory.all;
   }
 
+  get preState() {
+    return this._preState;
+  }
+
```
Comment (kevinbuhmann) [inline projects/demo/src/app/datagrid/detail/detail.ts:35]:
```suggestion
```

nit: no blank line because accessor pair

**Comment by kevinbuhmann** [comment]
Do we have guidance on when to disable vs. hide the open detail button? I'd like to review the documentation changes.

File: projects/demo/src/app/datagrid/detail/detail.html (hunk 13 above)
Comment (valentin-mladenov) [reply projects/demo/src/app/datagrid/detail/detail.html]: Corrected in commit cea931f

File: projects/angular/src/data/datagrid/datagrid-row.html (hunk 12 above)
Comment (valentin-mladenov) [reply projects/angular/src/data/datagrid/datagrid-row.html]: Corrected in commit cea931f

File: projects/angular/src/data/datagrid/datagrid-row.spec.ts [hunk 16]
```diff
@@ -543,6 +544,115 @@ export default function (): void {
...
+        const detailBody = document.querySelector('div.clr-dg-detail-body-wrapper') as HTMLElement;
+        expect(button.querySelector('cds-icon[shape^=angle-double][direction^=left]')).not.toBeNull();
+        expect(detailBody.innerText).toBe('' + context.testComponent.items[0].id);
+
+        button.click();
+        tick();
+        context.detectChanges();
+
+        expect(button.querySelector('cds-icon[shape^=angle-double][direction^=right]')).not.toBeNull();
+      }));
```
Comment (dtsanevmw) [inline projects/angular/src/data/datagrid/datagrid-row.spec.ts:597]: Not sure if we need that in the datagrid-row as this is already tested in detail unit tests and are part of the functionality there.

File: projects/angular/src/data/datagrid/datagrid-row.spec.ts [hunk 17]
```diff
@@ -543,6 +544,115 @@ export default function (): void {
...
+        context.testComponent.preState = context.testComponent.items[context.testComponent.disabledIndex];
+        tick();
+        context.detectChanges();
+
+        const button = context.clarityElement.querySelector('button.datagrid-detail-caret-button[disabled]');
+        expect(button).not.toBeNull();
+        expect(button.querySelector('cds-icon[shape^=angle-double][direction^=left]')).not.toBeNull();
+      }));
+
+      it('try open the hidden detail button', fakeAsync(function () {
```
Comment (dtsanevmw) [inline projects/angular/src/data/datagrid/datagrid-row.spec.ts:630]:
```suggestion
      it('check if detail button is hidden, fakeAsync(function () {
```

File: projects/angular/src/data/datagrid/datagrid-row.spec.ts [hunk 18]
```diff
@@ -543,6 +544,115 @@ export default function (): void {
...
+        context.testComponent.preState = context.testComponent.items[context.testComponent.hiddenIndex];
+        tick();
+        context.detectChanges();
+
+        const buttons = context.clarityElement.querySelectorAll('button.datagrid-detail-caret-button');
+        expect(buttons.length).toBe(1);
+
+        expect(context.clarityElement.querySelector('clr-dg-detail')).not.toBeNull();
+      }));
+    });
```
Comment (dtsanevmw) [inline projects/angular/src/data/datagrid/datagrid-row.spec.ts:655]: Not sure if we need that for the hidden functionality. We need to just verify if it's properly hiding the button and thats it.

File: projects/angular/src/data/datagrid/datagrid-row.spec.ts (hunk 16 above)
Comment (valentin-mladenov) [reply projects/angular/src/data/datagrid/datagrid-row.spec.ts:597]: It's not tested in `datagrid-detail.spec.ts`. It only tests detail pane, but not any open or close interaction through datagrid row.

Comment (dtsanevmw) [reply projects/angular/src/data/datagrid/datagrid-row.spec.ts:597]: It's okey then, thanks!

File: projects/angular/src/data/datagrid/datagrid-row.spec.ts (hunk 18 above)
Comment (valentin-mladenov) [reply projects/angular/src/data/datagrid/datagrid-row.spec.ts]: removed in commit 0ad320b

File: projects/angular/src/data/datagrid/datagrid-row.spec.ts (hunk 17 above)
Comment (valentin-mladenov) [reply projects/angular/src/data/datagrid/datagrid-row.spec.ts]: corrected in commit 0ad320b

**Review by dtsanevmw — APPROVED** [review]

**Pull request merged**